Initial draft of C FFI wrapper for filament

shartte opened the pull request:
This is an initial draft of a C wrapper for #139.

This will likely require major rework before it is merged, but I wanted to get the discussion going nevertheless.

Current the idea of API.h / cfilament.h is that the same symbol name is used in the C header, but is directly aliased to the C++ class if the header is included from the cpp file.

I am also not sure whether duplication of documentation makes sense here. The interface is most likely going to be used by writers of FFI wrappers for other languages that do not use the header directly. Actual users of C will likely have access to an actual C++ compiler and can link against the static library directly without use of this wrapper.

romainguy commented on cfilament/src/API.h:
```diff
@@ -0,0 +1,162 @@
+using FDepthPrepass = filament::View::DepthPrepass;
+
+namespace filament {
+  class Engine;
```
Code style rules: we use 4 spaces for normal indentation and 8 for continuation indentation.

romainguy commented on cfilament/src/Camera.cpp:
```diff
@@ -0,0 +1,128 @@
+using namespace filament;
+
+void Filament_Camera_SetProjectionFrustum(Camera *camera,
+                                          Camera::Projection projection,
```
Cf previous comment, we don't align parameters in our code style

romainguy commented on cfilament/src/Camera.cpp:
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
```
2018

romainguy commented on cfilament/src/API.h:
```diff
@@ -0,0 +1,162 @@
+/*
```
The name of the file is a bit generic, FilamentTypes or something like that would be better

romainguy commented on cfilament/src/API.h:
```diff
@@ -0,0 +1,162 @@
+ * limitations under the License.
+ */
+
+#pragma once
```
We currently don't use #pragma once, but regular header guards.

romainguy commented on cfilament/src/API.h:
```diff
@@ -0,0 +1,162 @@
+
+#pragma once
+
+#define CFILAMENT_SKIP_DECL
```
What's this define for?

shartte replied:
It disables the definition of the various opaque typedefs in the cfilament.h header so that they can be defined to the actual C++ types for compiling the wrapper implementations themselves. Alternatively, the entirety of API.h could be moved to a __cplusplus guard in the cfilament.h, if desired.

shartte commented:
I am gonna focus on fixing the compilation issues first.
Windows should work now, and I also removed --exclude-libs (which I added to truly hide all the C++ symbols) for the apple linker, which should fix OS X.
Could you please copy&paste the Android error, I can't see the Kokoro output and don't know how to compile for Android.

romainguy commented:
Here's the complete error log for Android:

```
[200/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Camera.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/Camera.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[201/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/VertexBuffer.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/VertexBuffer.cpp:22:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[202/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Colors.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/Colors.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[203/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Engine.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/Engine.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[204/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/EntityManager.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/EntityManager.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[205/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/IndirectLight.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/IndirectLight.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[206/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Fence.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/Fence.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[207/225] Building CXX object filament/test/CMakeFiles/filament_benchmark.dir/filament_benchmark.cpp.o
[208/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/IndexBuffer.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/IndexBuffer.cpp:23:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[209/225] Linking CXX executable filament/test/filament_benchmark
[210/225] Building CXX object filament/test/CMakeFiles/filament_calls.dir/filament_calls.cpp.o
[211/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/LightManager.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/LightManager.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[212/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Material.cpp.o
FAILED: cfilament/CMakeFiles/cfilament.dir/src/Material.cpp.o 
/tmpfs/src/github/filament/toolchains/Linux/arm-linux-androideabi-4.9/bin/arm-linux-androideabi-clang++   -DBUILD_CFILAMENT -Dcfilament_EXPORTS -I../../cfilament/include -I../../filament/include -I../../libs/math/include -I../../libs/utils/include -I../../third_party/robin-map/tnt/.. -I../../libs/filaflat/include -I../../libs/filabridge/include -I../../libs/image/include -mthumb -march=armv7-a -mcpu=cortex-a15 -mfloat-abi=softfp -mfpu=neon-vfpv4 -DANDROID -fPIE -std=c++14 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -fcolor-diagnostics -std=c++14 -O3 -DNDEBUG -fomit-frame-pointer -ffunction-sections -fdata-sections -fno-exceptions -fno-rtti -fPIC   -fvisibility=hidden -MD -MT cfilament/CMakeFiles/cfilament.dir/src/Material.cpp.o -MF cfilament/CMakeFiles/cfilament.dir/src/Material.cpp.o.d -o cfilament/CMakeFiles/cfilament.dir/src/Material.cpp.o -c /tmpfs/src/github/filament/cfilament/src/Material.cpp
In file included from /tmpfs/src/github/filament/cfilament/src/Material.cpp:21:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
/tmpfs/src/github/filament/cfilament/src/Material.cpp:93:1: error: static_assert failed "static size is wrong"
static_assert(sizeof(FParameter) == 16, "static size is wrong");
^             ~~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
[213/225] Linking CXX executable filament/test/filament_calls
[214/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/MaterialInstance.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/MaterialInstance.cpp:27:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[215/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/RenderableManager.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/RenderableManager.cpp:17:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[216/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Renderer.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/Renderer.cpp:21:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
[217/225] Building CXX object cfilament/CMakeFiles/cfilament.dir/src/Scene.cpp.o
In file included from /tmpfs/src/github/filament/cfilament/src/Scene.cpp:19:
In file included from /tmpfs/src/github/filament/cfilament/src/API.h:157:
../../cfilament/include/cfilament.h:153:14: warning: calling convention '__cdecl' ignored for this target [-Wignored-attributes]
typedef void(__cdecl *FFreeBufferFn)(void *buffer, size_t buffer_size, void *user);
             ^
1 warning generated.
ninja: build stopped: subcommand failed.
```

shartte commented:
Okay, I'll have to get ARM cross compilation working to truly work on that. I think some of my assumptions regarding structure packing, structure size and such might just be incorrect on ARM.

shartte commented:
Regarding the extern C question: Since all declarations in cfilament.h are inside an extern "C" block, that is inherited by the function implementations. I checked that the exported symbols are not mangled on both linux and windows, so it seems to be working as intended.

shartte commented:
I've updated a few of your requested changes, and also worked on a way of exporting a API JSON spec automatically. This might be useful to many writers of clients of this library to auto-generate their FFI code for this library (i.e. .NET P/Invoke declarations or similar). 
I am not sure whether there's a point in including the generation of this API JSON (Exaple here: https://gist.github.com/shartte/91d28b3aaab613394751f530dac14dd1) in this PR or not. Currently it's implemented by using libclang to parse the cfilament header, and extract structure layouts, enums, functions, etc. 
I could port this to something like Python, but it'd still depend on libclang to successfully extract all necessary info. Is filament built using clang on all platforms, or just some platforms? And could we assume availability of libclang, or rather not?

romainguy commented:
Build logs are now publicly accessible btw.

romainguy commented:
To answer your question: Filament is built with Clang on all platforms at the moment but there is clear demand to support other compilers as well. I don't know whether building with clang means libclang is available though. We could probably check in CMake?

romainguy commented:
We were just talking about this and since we need to maintain Java, JNI and Javascript versions of the API, being able to generate a JSON for the API would actually be extremely useful. We could write generators for the other languages from that JSON (we might want 2 JSONs, one "C-style" and one "C++-style" so we can recreate the appropriate class hierarchies etc. in Java for instance). `libclang` sounds like a great way to do this.

prideout commented:
@shartte we'd be interested in seeing the script you used to generate the JSON, we do have precedent for generating headers from Python (e.g. [bluevk-gen.py](https://github.com/google/filament/blob/b3d758f3b3fdf91b750a7561a1c729649cf4c1e8/libs/bluevk/bluevk-gen.py))

I think it would make sense to have a `libs/cfilament` that contains generated headers, it could also contain the Python script that was used to generate them.  I'd rather not add new folders to the repo root.

shartte commented on cfilament/src/View.cpp:
```diff
@@ -0,0 +1,112 @@
+    return view->getAntiAliasing();
+}
+
+void Filament_View_SetDynamicResolutionOptions(View *view, FDynamicResolutionOptions optionsIn) {
```
I have to be honest in that I didn't quite remember whether pure C allowed for const, heh. I'll check again and add const wherever it applies (in a lot of places).

shartte commented:
What I have right now is a .NET program that uses libclang to parse the actual wrapper API (cfilament.h) to produce an API JSON specification. Before I port that to something more suitable for inclusion (i.e. C++), I just wanted to make sure there's actually a point in doing so. 
Should I go ahead and port it to C++, and integrate it into the CMake build process in case libclang can be found on the host system?

prideout commented:
We'd actually be more interested in using the C++ headers as the "ground truth source" for the JSON, not the C headers.

We'd also be more interested in using C++ or Python for the JSON generator than .NET.

dlannan commented:
Just wondering if anyone tried using SWIG - seems to be an obvious choice here, then you dont need to be hand rolling C interfaces and you can automake Java/Python/C# interfaces?

prideout commented:
Good point, SWIG is an obvious choice.  However hand-rolling a generator from our fairly simple C++ API doesn't seem like overwhelming task, given the availability of [libclang for Python](https://pypi.org/project/clang/3.5/).  This would be friendly to IDE's, and would also allow us to more easily customize the results.

shartte commented:
So I think we should discuss what the way forward is. Libclang is actually an interesting case study because it solves the same problem. Provide a somewhat stable ABI to enable embedding libclang in other languages. In their case, they have handwritten the C wrapper around the C++ code in a similar way to this PR.

In general I think there are two ways to go here:

1) Provide a C interface, and publish a binary exposing this ABI alongside releases of filament. This will enable various languages to make use of filament without having to touch the filament source or native code at all (FFI approach). 

Languages I know of that can make use of such a library are: .NET Core (P/Invoke), Python (ctypes, cffi), Rust, Desktop Java (JNR)
Languages that may be able to use it, but suboptimally: JavaScript (node-ffi), Java (JNA)
For the suboptimal cases, it would likely still be possible to wrap the wrapper - so to speak - using auto-generated code.

2) Expect languages to provide their own wrapper code around the C++ interfaces directly, linking against the static library to produce a language specific filament DLL.

In this scenario, it might be helpful to provide a "spec" of the C++ API in a machine readable form. My concern here is that writing a spec format for a C++ API is potentially multiple orders of magnitude more complex than writing it for a C based API. Especially if one considers operator overloads and other concepts like it.

For the time being, I'll try to fix all concerns with the handwritten C wrapper until there's more guidance on how to proceed.

shartte commented:
I did start writing a Python/libclang based script to generate a specification of the C++ API.
I'll try to get to the level of the C wrapper proposed in this PR using that API spec and code generation. Once I am there, I'll open up another PR!

romainguy commented:
>My concern here is that writing a spec format for a C++ API is potentially multiple orders of magnitude more complex than writing it for a C based API. Especially if one considers operator overloads and other concepts like it.

I would generally agree but Filament's C++ API is fairly simple and does not rely on many advanced C++ language features. The few I can think of (templates for material parameters and libmath operator overloading) should be handled separately anyway. I would expect a generator based on the C++ API to come with a list of "exceptions" (Android does this with the Java API for OpenGL, most of it is generated, some of it is marked for manual implementation).

Our goal would be to avoid maintaining multiple API definitions, so it'd be nice if the C++ API was the spec.

prideout commented:
@shartte Thanks for the sneak peek into your Python scripts, I'm looking at them now :)

prideout commented:
Cool.  By the way, will the resulting headers compile using a pure C99 compiler, or are they merely ABI stable?

(I'm asking because I thought I saw C++ stuff in there like namespace.)

shartte commented:
In this PR, the header was dual-use. I tested the header from a client's perspective with "clang -xc -c" to make sure it'd compile in C mode as well. But it was also used by the cpp file with slightly different semantics. 
For the new code-generation approach I won't need the define hacks anymore, though.

prideout commented:
@shartte let us know if you need to set aside this work.

dlannan commented:
A note on SWIG - It will require alot of work to get SWIG working with it (Spent last three weeks trying to get something useful from it - on Win10 platform). There are too many template and other compiler issues that SWIG doesnt support well. Put simply, I do not think filament will port well to a C styled interface. You would be better writing a brand new one (interface) rather than attempt conversion/transformation (just my opinion).

shartte commented:
What's making my attempts at auto-generation a veritable nightmare are value types, move-semantics, and fuzzy ownership semantics of things like the entity-listener. I think the only benefit of such a complex auto-generation layer really is the maintenance of the different language glue code going forward, and the ability to more easily inherit documentation from the doxygen comments.

romainguy commented:
@shartte Which entity listener are you talking about? I know we have a few optional callbacks (for data upload) which will most like be a pain to handle. We should probably just have the ability to mark certain APIs as "manual" and require custom implementation per target language.

@dlannan The JNI layer is effectively a C API for Filament. Most of it is quite straightforward. But yeah, I wouldn't expect SWIG to be able to handle the API as it exists today.

shartte commented:
@romainguy It's not that I didn't look at the code, it's just not auto-detectable for the Python code parsing the header file.

dlannan commented:
@romainguy - while much of the code "appears" C like when you use SWIG it tries to work with some of the type management (the attributes for instance are not MSVC usable - with SWIG output) and numerous other semantics which causes a huge number of issues in SWIG to manually handle. So the problem becomes - do you spend a substantial amount of time making a bunch of specializations in SWIG templates, or do you spend similar time writing a wrapper interface. Im not interested in spending much more time in such an endeavor, because I think for me it would be a good 2-3 months to make a stable interface which I cannot afford right now. Its easier for me to select another rendering toolkit. Lazy? Maybe, but rendering is not my key target for my project - gk seems to be doing what I need, so I'll move on. Good luck with the conversion - I kind of wish there were more simplistic ways to tackle this.

romainguy commented:
@dlannan I was just saying that the API can be fairly easily wrapped under a C API (since that's what our JNI is). @shartte also has an interesting pure C API PR you can take a look at if interested. I'm definitely not advocating you spend time getting SWIG to work.

dlannan commented:
Oh. Btw. You can generate swig output with this:
[filament.zip](https://github.com/google/filament/files/2432259/filament.zip)
My main issues were on the compilation and integration side. I realized the above didnt make that very clear. Feel free to try using this in conversion. It did seem to generate reasonable python.

shartte commented:
@romainguy Any thought on my previous question? I'd like to continue, but before I do I'd like to confirm what acceptable options are for the JNI Code-Generator.

romainguy commented:
The JNI wrapper generator can be written in Java yes (or Kotlin). The only thing I don't want is to force a dependency on Java for developers who only want to build the native bits. Right now all the Java bits are optional and can be turned off.

shartte commented:
Alright I will get on that then. The JNI native library is a bit special because it actually needs to link against the Java libraries as well.
When you mentioned having to write a JavaScript interface, you probably meant Node.js ?

prideout commented:
Our JavaScript interface is for browsers, not nodejs.  For now we're using handrolling it using embind, which is working fairly well.
